File: CPP/explore/IsPalindrome.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
#include <cstdlib>
#include <vector>

struct ListNode
{
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(NULL) {}
};

class Solution
{
public:
  bool isPalindrome(ListNode *head)
  {

    if (head == NULL)
      return true;

    std::vector<int> forwardOrder;
    std::vector<int> reverseOrder;

    ListNode *currentNode = head;
    int counter = 0;

    while (currentNode != NULL)
    {
      forwardOrder.push_back(currentNode->val);
      counter += 1;
      currentNode = currentNode->next;
    }

    for (int i = counter - 1; i >= 0; i--)
    {
      reverseOrder.push_back(forwardOrder[i]);
    }

    for (int j = 0; j < counter; j++)
    {
      if (forwardOrder[j] != reverseOrder[j])
        return false;
    }

    return true;
  }
};

File: JS/explore/IsPalindrome.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function(head) {
	if (head === null) return true;

	// This method just stores the values of the linked list in the array since access is much easier.
	let forwardOrder = new Array();

	let currentNode = head;
	while (currentNode !== null) {
		forwardOrder.push(currentNode.val);
		currentNode = currentNode.next;
	}

	// The JS reverse method is destructive, need to make a new instance.
	let reverseOrder = forwardOrder.slice().reverse();

	console.log(forwardOrder, reverseOrder);

	for (let i = 0; i < forwardOrder.length; i++) {
		if (forwardOrder[i] !== reverseOrder[i]) return false;
	}

	return true;
};
